include/Gaffer/Signals.inl

```diff
@@ -0,0 +1,420 @@
+template<typename SlotFunctor>
+Connection Signal<Result( Args... ), Combiner>::connect( const SlotFunctor &slot )
+{
+	return connectInternal( slot, /* front = */ false );
+}
+
+template<typename Result, typename... Args, typename Combiner>
+template<typename SlotFunctor>
+Connection Signal<Result( Args... ), Combiner>::connectFront( const SlotFunctor &slot )
+{
+	return connectInternal( slot, /* front = */ true );
+}
+
+template<typename Result, typename... Args, typename Combiner>
+template<typename SlotFunctor>
+Connection Signal<Result( Args... ), Combiner>::connectInternal( const SlotFunctor &slot, bool front )
+{
+	if( !m_lastSlot )
+	{
+		assert( !m_firstSlot );
+		m_lastSlot = new Slot( m_firstSlot );
+	}
+
+	Private::SlotBase::Ptr s = new Slot(
+		front ? m_firstSlot : *(m_lastSlot->previous), slot
+	);
+
+	const Connection result = Connection( s );
+	Trackable::trackConnection( slot, result );
+	return result;
+}
+
+template<typename Result, typename... Args, typename Combiner>
+template<typename SlotFunctor>
+void Signal<Result( Args... ), Combiner>::disconnect( const SlotFunctor &slotFunctor )
+{
+	Private::SlotBase::Ptr slot = m_firstSlot;
+	while( slot != m_lastSlot )
+	{
+		if( static_cast<Slot *>( slot.get() )->function == slotFunctor )
+		{
+			slot->disconnect();
+		}
+		slot = slot->next;
+	}
+};
+
+template<typename Result, typename... Args, typename Combiner>
+Result Signal<Result( Args... ), Combiner>::operator() ( Args... args ) const
+{
+	ArgsTuple argsTuple( args... ); /// \todo : Capture by reference? Or forward_as_tuple?
+	return m_combiner(
+		SlotCallIterator( m_firstSlot, m_lastSlot, argsTuple ),
+		SlotCallIterator( m_lastSlot, m_lastSlot, argsTuple )
+	);
+}
+
+template<typename Result, typename... Args, typename Combiner>
+void Signal<Result( Args... ), Combiner>::disconnectAllSlots()
+{
+	while( m_firstSlot != m_lastSlot )
+	{
+		m_firstSlot->disconnect();
+	}
+	assert( m_firstSlot == m_lastSlot );
+};
+
+template<typename Result, typename... Args, typename Combiner>
+size_t Signal<Result( Args... ), Combiner>::numSlots() const
+{
+	size_t result = 0;
+	const Private::SlotBase *s = m_firstSlot.get();
+	while( s != m_lastSlot.get() )
+	{
+		result++;
+		s = s->next.get();
+	}
+	return result;
+};
+
+template<typename Result, typename... Args, typename Combiner>
+bool Signal<Result( Args... ), Combiner>::empty() const
+{
+	return m_firstSlot == m_lastSlot;
+}
+
+//////////////////////////////////////////////////////////////////////////
+// Signal::Slot
+//////////////////////////////////////////////////////////////////////////
+
+template<typename Result, typename... Args, typename Combiner>
+struct Signal<Result( Args... ), Combiner>::Slot : public Private::SlotBase
+{
+
+	/// \todo We'd like to use `std::function` here, but it doesn't support
+	/// equality comparison, which makes implementing
+	/// `Signal::disconnect( const SlotFunctor & )` hard (impossible?) to
+	/// implement.
```

Comment:
I was trying a little to understand what's happening here.  I guess the alternative would be to require some sort of explicit handle to allow removing later?  When looking up what boost::function does, I just found this "https://www.boost.org/doc/libs/1_78_0/doc/html/function/faq.html", which appears to be current, and says that you can't use equality comparison on boost functions, which didn't reduce my confusion.

Reply:
> I guess the alternative would be to require some sort of explicit handle to allow removing later?

Correct, and the `Connection` class provides that handle. But we have code that avoids storing the connection by calling `disconnect( functor )`, and I want to continue to support that, at least for now.

> I just found this "https://www.boost.org/doc/libs/1_78_0/doc/html/function/faq.html", which appears to be current, and says that you can't use equality comparison on boost functions, which didn't reduce my confusion.

That confused me too! And it's also in the docs for the versions we're currently using. What's not very clear is that `boost::function == boost::function` _isn't_ supported, but `function == arbitraryFunctorYouCouldConstructAFunctionFrom` _is_ supported. My vague understanding of this is that some type information (about the functor) is lost once the `boost::function` has been constructed. But the `== arbitaryFunctor` case provides that type information again, so is possible.

`std::function` does provide `target_type` and `target` queries which I suspect could be used to compare to a functor, but I haven't got my head around it enough to implement that yet.